# xls → csv 변환하기
유튜브: https://youtu.be/38B_Q0OdYEs

In [110]:
import glob
import xlrd
import pandas as pd
import datetime

In [97]:
xls = glob.glob('*.xls')
xls

['디알텍.xls']

In [98]:
xls[0]

'디알텍.xls'

In [99]:
wb = xlrd.open_workbook(xls[0], encoding_override='cp1252')
ws = wb.sheet_by_index(0)

In [26]:
ws.ncols

16

In [27]:
ws.nrows

569

In [29]:
ws.col_values(0)[:5]

['날짜', '2022/05/04', '2022/05/04', '2022/05/04', '2022/05/04']

### 모든 row의 length가 같은지 확인하기
모두 569개로 똑같다.

In [31]:
for i in range(ws.ncols):
    print(len(ws.col_values(i)))

569
569
569
569
569
569
569
569
569
569
569
569
569
569
569
569


### 마지막 칼럼의 마지막 5개 row data는 빈칸이지만 ''로 지정되어 있다.
따라서 데이터가 존재하든 존재하지 않든 모든 칼럼의 길이는 같다.

In [37]:
ws.col_values(15)[-5:]

['', '', '', '', '']

### row의 첫줄은 칼럼의 제목이다.

In [38]:
ws.row_values(0)

['날짜',
 '시간',
 '시가',
 '고가',
 '저가',
 '종가',
 '종가 단순 5',
 10.0,
 20.0,
 60.0,
 120.0,
 '거래량',
 '단순 5',
 20.0,
 60.0,
 120.0]

### 데이터프레임 칼럼에 사용할 제목으로 할당해준다.

In [41]:
titles = ws.row_values(0)
titles[:5]

['날짜', '시간', '시가', '고가', '저가']

### dataframe 만들기

In [79]:
pieces = []
for i in range(ws.ncols):
    pieces.append(ws.col_values(i)[1:])

In [80]:
df = pd.DataFrame(pieces, index=titles).T
df.head()

날짜        시간    시가    고가    저가    종가 종가 단순 5  10.0    20.0  \
0  2022/05/04  09:35:00  5580  5770  5580  5670    5578  5504  5400.5   
1  2022/05/04  09:34:00  5540  5650  5530  5600    5564  5469    5381   
2  2022/05/04  09:33:00  5570  5580  5510  5550    5538  5438  5363.5   
3  2022/05/04  09:32:00  5530  5610  5500  5570    5508  5410  5348.5   
4  2022/05/04  09:31:00  5600  5650  5430  5500    5460  5380    5333   

      60.0    120.0     거래량     단순 5     20.0     60.0    120.0  
0  5298.67  5264.83  105145  51909.8    20464  9322.48  4935.09  
1     5291  5261.08   45218  41484.8    15265  7575.08  4063.04  
2  5284.33  5257.92   21372  41667.6  13193.1  6825.68  3696.22  
3  5278.67   5255.5   30320  39487.8    12186  6478.58  3518.13  
4   5272.5   5252.5   57494  33907.4  11065.4  6059.75  3267.13

In [82]:
name = xls[0].split('.')[0]
name

'대한전선'

In [85]:
df.to_csv(name+'.csv', index=False)

# 함수(function) 만들기

In [119]:
xls = glob.glob('*.xls')

for x in xls:
    wb = xlrd.open_workbook(x, encoding_override='cp1252')
    ws = wb.sheet_by_index(0)
    
    pieces = []
    for i in range(ws.ncols):
        pieces.append(ws.col_values(i)[1:])

    df = pd.DataFrame(pieces, index=titles).T
    
    df['날짜'] = pd.to_datetime(df['날짜']).apply(lambda x : x.date())
    target_date = datetime.datetime(2022,5,4).date()
    df = df[df['날짜'] == target_date]
    
    name = x.split('.')[0]
    df.to_csv(name+'.csv', index=False)